File: core/core-domain/src/main/java/com/coffee/gu/stamp/StampRepository.java
package com.coffee.gu.stamp;

import java.time.LocalDateTime;
import java.util.Collection;
import java.util.List;

public interface StampRepository {
    Stamp save(Stamp stamp);
    List<Stamp> saveAll(Collection<Stamp> stamps);
    Long countAvailableStamps(String principalKey, LocalDateTime now);
    List<Stamp> getAvailableStamps(String principalKey, LocalDateTime now, int pageNumber, int pageSize);
    Long countExpiringStamps(String principalKey, LocalDateTime now, LocalDateTime nDaysLater);
    List<Stamp> findByOrderKey(String orderKey);
}

File: core/core-domain/src/main/kotlin/com/coffee/gu/stamp/StampRepository.kt
package com.coffee.gu.stamp

import java.time.LocalDateTime

interface StampRepository {
    fun save(stamp: Stamp): Stamp
    fun saveAll(stamps: Collection<Stamp>): List<Stamp>
    fun countAvailableStamps(principalKey: String, now: LocalDateTime): Long
    fun getAvailableStamps(principalKey: String, now: LocalDateTime, pageNumber: Int, pageSize: Int): List<Stamp>
    fun countExpiringStamps(principalKey: String, now: LocalDateTime, nDaysLater: LocalDateTime): Long
    fun findByOrderKey(orderKey: String): List<Stamp>
}
